src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -0,0 +1,160 @@
+require 'puppet/server'
+
+module Puppet
+  module Server
+    class Compiler
+
+      def initialize
+        set_server_facts
+      end
+
+      def compile(request_data)
+        processed_hash = convert_java_args_to_ruby(request_data)
+
+        node = create_node(processed_hash)
+
+        catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
+        catalog.to_data_hash
+      end
+
+      private
+
+      def create_node(request_data)
+        # We need an environment to talk to PDB
+        request_data['environment'] ||= 'production'
+
+        facts, trusted_facts = process_facts(request_data)
+        node_params = { facts: facts,
+                        # TODO: fetch environment from classifier
+                        environment: request_data['environment'],
+                        # data added to the node object and exposed in manifests as
+                        # top-level vars. Maybe related to class params??
+                        # Can these also come from the classifier?
+                        parameters: request_data['parameters'],
+                        # TODO: fetch classes from classifier
+                        classes: request_data['classes'] }
+
+        node = Puppet::Node.new(request_data['certname'], node_params)
+        # Merges facts into the node parameters.
+        # Ensures that facts will be surfaced as top-scope variables,
+        # along with other node parameters.
+        node.merge(facts.values)
+        node.trusted_data = trusted_facts
+        node.add_server_facts(@server_facts)
+        node
+      end
+
+      def convert_java_args_to_ruby(hash)
+        Hash[hash.collect do |key, value|
+            # Stolen and modified from params_to_ruby in handler.rb
+            newkey = key.to_s
+            # Java::ClojureLang::Keywords retain their leading colon when
+            # converted to a string
+            newkey.delete_prefix!(':')
```

Comment:
I think this can be fixed later, but we should probably use [stringify-keys](https://clojuredocs.org/clojure.walk/stringify-keys) on the clojure side prior to calling compileCatalog.

Reply:
I can add that. That seems maybe better.

Reply:
Ended up instead changing the JSON parsing to not turn these keys into keywords in the first place.